pallets/oracle-data-collection/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+			key: T::OracleKey,
+		},
+		UpdatedFeeders {
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+			feeders: BoundedVec<T::AccountId, T::MaxFeedersPerKey>,
+		},
+		UpdatedCollection {
+			collection_id: T::CollectionId,
+			keys_updated: u32,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The account who trigger the action is not a collection admin.
+		IsNotAdmin,
+
+		/// The key is not in the collection.
+		KeyNotInCollection,
+
+		/// Collection size reached
+		MaxCollectionSize,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Update the feeders associated to an specific key.
+		///
+		/// NOTE: In the future this should be go though change guard.
+		/// By now we trust in the admin to choose honest feeders without
+		/// invertor supervision
+		#[pallet::weight(1_000_000)]
+		#[pallet::call_index(2)]
+		pub fn update_feeders(
+			origin: OriginFor<T>,
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+			feeders: BoundedVec<T::AccountId, T::MaxFeedersPerKey>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				T::IsAdmin::check((who, collection_id)),
+				Error::<T>::IsNotAdmin
+			);
+
+			Self::mutate_and_remove_if_clean(&collection_id, &key, |info| {
+				info.feeders = feeders.clone();
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::<T>::UpdatedFeeders {
+				collection_id: collection_id,
+				key: key,
+				feeders,
+			});
+
+			Ok(())
+		}
+
+		/// Update the collection, doing the aggregation for each key in the
+		/// process
+		#[pallet::weight(1_000_000)]
+		#[pallet::call_index(3)]
+		pub fn update_collection(
+			origin: OriginFor<T>,
+			collection_id: T::CollectionId,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			let values = Keys::<T>::iter_key_prefix(&collection_id)
+				.map(|key| Self::get(&key, &collection_id).map(|value| (key, value)))
+				.collect::<Result<Vec<_>, _>>()?;
+
+			let len = values.len();
+
+			let collection = BoundedBTreeMap::try_from(BTreeMap::from_iter(values))
+				.map_err(|()| Error::<T>::MaxCollectionSize)?;
+
+			Collection::<T>::insert(collection_id, collection);
+
+			Self::deposit_event(Event::<T>::UpdatedCollection {
+				collection_id: collection_id,
+				keys_updated: len as u32,
+			});
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> DataRegistry<T::OracleKey, T::CollectionId> for Pallet<T> {
+		type Collection = CachedCollection<T>;
+		type Data = (T::OracleValue, T::Moment);
+		#[cfg(feature = "runtime-benchmarks")]
+		type MaxCollectionSize = T::MaxCollectionSize;
+
+		fn get(
+			key: &T::OracleKey,
+			collection_id: &T::CollectionId,
+		) -> Result<Self::Data, DispatchError> {
+			let key_info = Keys::<T>::get(collection_id, key);
+			let fed_values = key_info
+				.feeders
+				.into_iter()
+				.map(|feeder| T::OracleProvider::get(&(feeder, *collection_id), key))
+				.collect::<Result<Vec<_>, _>>()?;
+
+			let (mut values, mut timestamps): (Vec<_>, Vec<_>) = fed_values.into_iter().unzip();
```

Comment:
Am I understanding this correctly that we fetch the values here instead of using the cached ones?

Reply:
Yes. `The OracleProvider::get()` is who fetches. I'm open to using the cached ones, but we would need to read the entire collection for just one value, which could be a PoV issue. What's your point of view regarding this?

Reply:
> What's your point of view regarding this?

Sorry, I misunderstood. So this is jsut for getting a single value and then if you want a collection you take the cached one? This is great! And agree with the PoV issue otherwise.